### PR TITLE
Adding investigation and dataset card views for DataGateway search #420

### LINKED ISSUE #420: UI improvement: add toggle card in search results
**Description**:
Investigate if easy to add toggle to card view in the search plugin

**Acceptance criteria**:
- [ ] toggle card view in search

### PR DESCRIPTION
**Pull request by sam-glendenning**
## Description
This adds card views when searching in DataGateway in investigations and datasets (datafiles do not have card views). A toggle allows the user to switch between table and card view, with the default being the table view. 

## Testing instructions
Still a work in progress

- [ ] Review code
- [ ] Check Actions build
- [ ] Review changes to test coverage
- [ ] {more steps here}

## Agile board tracking
Closes #420

### CODE REVIEW

**Comment by sam-glendenning** [comment]
Screenshot of the work done so far:

image

**Comment by GoelBiju** [comment]
> Screenshot of the work done so far:
> 
> image

This looks really nice @sam-glendenning! 😃

**Comment by sam-glendenning** [comment]
@GoelBiju @louise-davies 

Didn't manage to get this totally finished. It's 90% there and all the functionality is there, just needs a bit more done.
- `investigationSearchCardView` component needs unit testing (much of this can be copied from `datasetSearchCardView` and changed accordingly)
- e2e tests need to be added/changed in accordance with new functionality
- small UI bug when on card view where, behind the first card, you can see the outline of where the table component would belong. This is still being rendered due to the `Tabs` being displayed. This will need to be fixed in some way
- small UI bug when changing between table and card view, the component gets slightly bigger and smaller. This is because the table is confined to a box with an independent scroll bar whereas the card view expands the page itself. Not sure what solution could be applied here
- cards do not currently have an Add to Cart or Download button (code for this is commented out in places so shouldn't be hard to add in if wanted). At the time, wasn't sure if I wanted to keep it in but on reflection I think it makes sense to have those buttons in

This PR looks massive when you see how many new lines are being added but the majority of that is one or two new snapshots thousands of lines in length. Other components in dg search suffer from long snapshots as well, they may be able to be trimmed down at some point.

**Comment by GoelBiju** [comment]
> @GoelBiju @louise-davies
> 
> Didn't manage to get this totally finished. It's 90% there and all the functionality is there, just needs a bit more done.
> 
> * `investigationSearchCardView` component needs unit testing (much of this can be copied from `datasetSearchCardView` and changed accordingly)
> * e2e tests need to be added/changed in accordance with new functionality
> * small UI bug when on card view where, behind the first card, you can see the outline of where the table component would belong. This is still being rendered due to the `Tabs` being displayed. This will need to be fixed in some way
> * small UI bug when changing between table and card view, the component gets slightly bigger and smaller. This is because the table is confined to a box with an independent scroll bar whereas the card view expands the page itself. Not sure what solution could be applied here
> * cards do not currently have an Add to Cart or Download button (code for this is commented out in places so shouldn't be hard to add in if wanted). At the time, wasn't sure if I wanted to keep it in but on reflection I think it makes sense to have those buttons in
> 
> This PR looks massive when you see how many new lines are being added but the majority of that is one or two new snapshots thousands of lines in length. Other components in dg search suffer from long snapshots as well, they may be able to be trimmed down at some point.

No problem @sam-glendenning, it looks great so far anyway. I'll have a look at what I can add to it as well!

**Comment by GoelBiju** [comment]
> @GoelBiju @louise-davies
> 
> Didn't manage to get this totally finished. It's 90% there and all the functionality is there, just needs a bit more done.
> 
> * `investigationSearchCardView` component needs unit testing (much of this can be copied from `datasetSearchCardView` and changed accordingly)
> * e2e tests need to be added/changed in accordance with new functionality
> * small UI bug when on card view where, behind the first card, you can see the outline of where the table component would belong. This is still being rendered due to the `Tabs` being displayed. This will need to be fixed in some way
> * small UI bug when changing between table and card view, the component gets slightly bigger and smaller. This is because the table is confined to a box with an independent scroll bar whereas the card view expands the page itself. Not sure what solution could be applied here
> * cards do not currently have an Add to Cart or Download button (code for this is commented out in places so shouldn't be hard to add in if wanted). At the time, wasn't sure if I wanted to keep it in but on reflection I think it makes sense to have those buttons in
> 
> This PR looks massive when you see how many new lines are being added but the majority of that is one or two new snapshots thousands of lines in length. Other components in dg search suffer from long snapshots as well, they may be able to be trimmed down at some point.

Just to record what is being completed:

- [ ] `investigationSearchCardView` unit tests,
- [ ] e2e tests for new functionality,
- [ ] outline of table underneath card view,
- [ ] changing between table and card view gets larger due to scroll bar appearing,
- [ ] add "Add to Cart" or "Download" button to cards.

**Comment by agbeltran** [comment]
@joelvdavies please discuss with @GoelBiju or @louise-davies about this one

**Comment by joelvdavies** [comment]
@louise-davies Should this have  'add to cart' and 'download' buttons or just the download. At the moment the cart is not displayed on the page unlike on dataview.

**Comment by louise-davies** [comment]
It should still have an add to cart button - perhaps later we'll add a quick "take me to cart" button like dataview

File: packages/datagateway-search/src/table/datasetSearchTable.component.tsx [hunk 1]
```diff
@@ -168,7 +168,7 @@ const DatasetSearchTable = (props: DatasetTableProps): React.ReactElement => {
       return linkType === 'investigation'
         ? tableLink(
             `/browse/proposal/${datasetData.investigation.name}/investigation/${datasetData.investigation.id}/dataset`,
-            datasetData.investigation.title
+            datasetData.investigation.name
```
Comment (louise-davies) [inline packages/datagateway-search/src/table/datasetSearchTable.component.tsx:171]: Why was this change made? This needs to be switched back as the original was correct (not just for this line, but all lines like this in this file)

**Review by louise-davies — CHANGES_REQUESTED** [review]
Minor issue but if you switch to datafile when card view is selected, the button gets greyed out but it still says "display as table" - this should either change to say "display as cards", or perhaps remove the button altogether (which is what dataview does). Very minor visual bug though!

File: packages/datagateway-search/cypress/integration/searchPageContainer.spec.ts [hunk 2]
```diff
@@ -1,13 +1,122 @@
...
+  //     .click();
+  //   cy.get('#card').contains('Someone statement Republican plan watch.');
+
+  //   cy.get('[aria-label="Go to first page"]', { timeout: 10000 })
+  //     .first()
+  //     .click();
+  //   cy.get('#card').contains(
+  //     'Including spend increase ability music skill former.'
+  //   );
+  // });
```
[the remark this answers is not in the document]
Comment (joelvdavies) [reply packages/datagateway-search/cypress/integration/searchPageContainer.spec.ts]: I have replaced it with it.skip.

File: packages/datagateway-search/src/table/datasetSearchTable.component.tsx (hunk 1 above)
Comment (joelvdavies) [reply packages/datagateway-search/src/table/datasetSearchTable.component.tsx]: I just had a look at the master and corrected all of them.

Comment (sam-glendenning) [reply packages/datagateway-search/src/table/datasetSearchTable.component.tsx]: I _believe_ this change was actually done by me (but don't quote me on that, check the git blame). My reasoning was that the title was much longer and wordier so didn't fit easily into the table column compared to the name which was always something like `INVESTIGATION 16`. As Louise says, perhaps this wasn't the right idea and wouldn't work well with production data. So I would suggest following her advice.

File: packages/datagateway-search/src/card/datasetSearchCardView.component.tsx [hunk 3]
```diff
@@ -0,0 +1,311 @@
...
+              ? sizeQueries[index]
+              : datasetCountQueries[index];
+          return formatCountOrSize(query, hierarchy === 'isis');
+        },
+        disableSort: true,
+      },
+      {
+        icon: Fingerprint,
+        label: t('datasets.investigation'),
+        dataKey: 'investigation.name',
```
Comment (louise-davies) [inline packages/datagateway-search/src/card/datasetSearchCardView.component.tsx:236]: This should match the table and I think this needs to be `title` right (due to my previous comment)?

File: packages/datagateway-search/src/searchPageContainer.component.tsx [hunk 4]
```diff
@@ -106,7 +106,11 @@ const ViewButton = (props: {
         onClick={() => props.handleButtonChange()}
         disabled={props.disabled}
       >
-        {props.viewCards ? t('app.view_table') : t('app.view_cards')}
+        {props.viewCards
```
[the remark this answers is not in the document]
Comment (joelvdavies) [reply packages/datagateway-search/src/searchPageContainer.component.tsx:109]: I agree, I have now changed this.

**Review by louise-davies — APPROVED** [review]

**Pull request merged**